subsys/infinispan/src/main/java/org/commonjava/indy/subsys/infinispan/CacheProducer.java

```diff
@@ -75,20 +80,47 @@
     @Inject
     private IndyMetricsConfig metricsConfig;
 
-    private Map<String, CacheHandle> caches = new ConcurrentHashMap<>();
+    private Map<String, BasicCacheHandle> caches = new ConcurrentHashMap<>(); // hold embedded and remote caches
 
     protected CacheProducer()
     {
     }
 
-    public CacheProducer( IndyConfiguration indyConfiguration, EmbeddedCacheManager cacheManager )
+    public CacheProducer( IndyConfiguration indyConfiguration, EmbeddedCacheManager cacheManager, ISPNRemoteConfiguration remoteConfiguration )
     {
         this.indyConfiguration = indyConfiguration;
         this.cacheManager = cacheManager;
+        this.remoteConfiguration = remoteConfiguration;
+        if ( remoteConfiguration != null )
+        {
+            startRemoteManager();
+        }
     }
 
     @PostConstruct
     public void start()
+    {
+        startRemoteManager();
+        startEmbeddedManager();
+    }
+
+    private RemoteCacheManager remoteCacheManager;
+
+    private void startRemoteManager()
+    {
+        if ( !remoteConfiguration.isEnabled() )
+        {
+            logger.info( "Infinispan remote configuration not enabled. Skip." );
+            return;
+        }
+
+        ConfigurationBuilder builder = new ConfigurationBuilder();
+        builder.addServer().host( remoteConfiguration.getRemoteServer() ).port( remoteConfiguration.getHotrodPort() );
```

Comment:
Will this need to be a list of remote servers in order to handle clustering? What if the one we have configured goes down but the cluster is still up?

Reply:
It seems we just add more servers and it will balance. We have two ways to run ISPN cluster. The standalone mode of multiple servers; and domain mode. The standalone mode is easy to set up and start but hard to manage. Domain mode is better on management but need more steps to set it up. This is the area we have little experience now.